src/ape-together/ApeRebalanceExtension.sol

```diff
@@ -0,0 +1,265 @@
+        currentEpochStart = block.timestamp;
+        for (uint256 i = 0; i < possibleComponents.length; i++) {
+            votes[possibleComponents[i]] = 0;
+        }
+        delete possibleComponents;
+    }
+
+    /**
+     * Overrides the original rebalance function from GIMExtension. Always reverts.
+     */
+    function startRebalanceWithUnits(
+        address[] memory /* _components */,
+        uint256[] memory /* _targetUnits */,
+        uint256 /* _positionMultiplier */
+    )
+        external
+        override
+    {
+        revert("only democratically elected shitcoins allowed");
+    }
+
+    /**
+     * Fetches the current top voted components and weights. When the rebalance begins,
+     * it will set the weights to be identical to the weights given by this function. The
+     * weights are measured as the percentage of the toal index value, not the unit amounts.
+     *
+     * @return addres[]     top voted on components
+     * @return uint256[]    components weights (not units) as per the vote
+     */
+    function getWeights() external view returns (address[] memory, uint256[] memory) {
+        return _getWeights();
+    }
+
+    /* ========= Internal Functions ========== */
+
+    /**
+     * Fetches the total number of votes of a user. This function allows for an address to
+     * hold multiple OwlNFTs.
+     *
+     * @param _voter        address of voter to check votes for
+     * @return uint256      number of votes that the address has
+     */
+    function _getVotes(address _voter) internal view returns (uint256) {
+        uint256 bal = owlNft.balanceOf(_voter);
+        
+        uint256 totalVotes;
+        for (uint256 i = 0; i < bal; i++) {
+            uint256 id = owlNft.tokenOfOwnerByIndex(_voter, i);
+            totalVotes = totalVotes.add(owlNft.getVotes(id));
+        }
+
+        return totalVotes;
+    }
+
+    /**
+     * Checks whether the current epoch has ended.
+     *
+     * @return bool     whether current epoch has ended
+     */
+    function _isEpochOver() internal view returns (bool) {
+        return block.timestamp >= currentEpochStart.add(epochLength);
+    }
+
+    /**
+     * Fetches the current top voted components and weights. When the rebalance begins,
+     * it will set the weights to be identical to the weights given by this function. The
+     * weights are measured as the percentage of the toal index value, not the unit amounts.
+     *
+     * @return components   top voted on components
+     * @return weights      components weights (not units) as per the vote
+     */
+    function _getWeights() internal view returns (address[] memory components, uint256[] memory weights) {
+        
+        address[] memory possibleLeft = possibleComponents;
+        uint256 numComponents = Math.min(maxComponents, possibleComponents.length);
+        components = new address[](numComponents);
+
+        // forgive me father for I have sinned with this selection sort
```

Comment:
Since the outer loop is capped it to `maxComponents` (which I think we probably will want to set at ~10) it takes only 10*n iterations which is *technically* O(n).

![image](https://user-images.githubusercontent.com/16858330/139953943-85c89908-0dd8-4814-bc44-23026a04d367.png)